lib/charms/alertmanager_k8s/v0/alertmanager.py

```diff
@@ -0,0 +1,146 @@
+LIBAPI = 0  # Must match the major version in the import path.
+LIBPATCH = 1  # The current patch version. Must be updated when changing.
+
+logger = logging.getLogger(__name__)
+
+
+class AlertmanagerConsumer(ConsumerBase):
+    """A "consumer" handler to be used by charms that relate to Alertmanager.
+
+    This consumer auto-registers relation events on behalf of the user and communicates information
+    directly via `_stored`
+    Every change in the alertmanager cluster emits an 'on.available' event that the consumer charm
+    can register and handle, for example:
+
+        self.framework.observe(self.alertmanager_lib.on.available,
+                               self._on_alertmanager_cluster_changed)
+
+    The updated alertmanager cluster can then be obtained via the
+
+    This consumer library expect the consumer charm to register the `get_cluster_info` method.
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): from consumer's metadata.yaml
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+
+    _stored: StoredState
+
+    def __init__(self, charm: CharmBase, relation_name: str, consumes: dict, multi: bool = False):
+        super().__init__(charm, relation_name, consumes, multi)
+        self.charm = charm
+        self._consumer_relation_name = relation_name  # from consumer's metadata.yaml
+        # self._provider_relation_name = "alerting"  # from alertmanager's metadata.yaml
+
+        self.framework.observe(
+            self.charm.on[self._consumer_relation_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            self.charm.on[self._consumer_relation_name].relation_departed,
+            self._on_relation_departed,
+        )
+        self.framework.observe(
+            self.charm.on[self._consumer_relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self._stored.set_default(alertmanagers={})
+
+    def _on_relation_changed(self, event: ops.charm.RelationChangedEvent):
+        """This hook stores locally the address of the newly-joined alertmanager.
+        This is needed for consumers such as prometheus, which should be aware of all alertmanager
+        instances.
+        """
+        if event.unit:  # event.unit may be `None` in the case of app data change
+            # Save locally the public IP address of the alertmanager unit
+            if address := event.relation.data[event.unit].get("public_address"):
+                # TODO consider storing in unit data instead of StoredState
+                self._stored.alertmanagers[event.unit.name] = address
+
+                # inform consumer about the change
+                # manually emitting the "available" event until scale-up/down "ready" flag is fixed
+                self.on.available.emit()
```

Comment:
This is a very good argument in favor of new custom event (e.g. `ClusterChangedEvent`), because if I try to stick to OF events then the consumer charms would have to register available, invalid, etc.

Reply:
Replaced with a custom event.